### ISSUE TITLE
Freeze/Delay occurs when uploading blobs to Iot Edge Blob Storage Module

### ISSUE THREAD

**Issue by kc-95-scg** [Azure/azure-sdk-for-cpp#2647]
**Describe the bug**
Freeze/Delay occurs when uploading blobs to Iot Edge Blob Storage Module.
The delay/freeze is not random, and it happens exactly every 100 blob operations (ie UploadFrom( ), SetMetadata( ), etc) to the IoT Edge Blob Storage module.

This is an example of the delay when uploading blobs to this module.
![Screenshot from 2021-07-16 10-09-02](https://user-images.githubusercontent.com/78844564/126034058-de219220-313e-44d5-bb97-a7491359f42a.png)

The delay **only occurs** when we are using this Azure SDK for CPP to upload to Iot Edge Blob Storage Module, (which is running locally on an Iot Edge device.)

The delay **doesnt occur** when using the same code to upload to a cloud storage account, or the development azurite container.
The delay **doesnt occur** when using another storage SDK (tested with python and javascript) to the Iot Edge Blob Storage Module.

***Exception or Stack Trace***
No errors are thrown when the freeze/delay occurs.
The program just hangs and awaits for the blob operation to complete.
No error information in the logs of the blob storage module. 
No error information in the logs of the Iot Edge edgeHub & edgeAgent module.

**To Reproduce**

Create an Iotedge device and deploy the iotedge blob storage module.
The following parameters were used in the deployment manifest
```
"modules": {
        "local-azure-storage": {
        "version": "1.0",
        "type": "docker",
        "status": "running",
        "restartPolicy": "always",
        "startupOrder": 1,
        "settings": {
            "image": "mcr.microsoft.com/azure-blob-storage:1.4.1-linux-amd64",
            "createOptions": {
            "Env": [
                "LOCAL_STORAGE_ACCOUNT_NAME=localstorageacc",
                "LOCAL_STORAGE_ACCOUNT_KEY=password"
            ],
            "HostConfig": {
                "LogConfig": {
                "Type": "json-file",
                "Config": {
                    "max-size": "10m",
                    "max-file": "3"
                    }
                },
                "Binds": [
                    "/localblob:/blobroot"
                ],
                "PortBindings": {
                    "11002/tcp": [
                        {
                            "HostPort": "11002"
                        }
                    ]
                }
            }
            }
        }
    }
}

"local-azure-storage": {
    "properties.desired": {
    "deviceAutoDeleteProperties": {
        "deleteOn": false,
        "deleteAfterMinutes": 3600,
        "retainWhileUploading": false
    },
    "deviceToCloudUploadProperties": {
        "uploadOn": false,
        "uploadOrder": "OldestFirst",
        "cloudStorageConnectionString": "",
        "storageContainersForUpload": {
            }
        },
        "deleteAfterUpload": true
    }
    }
},
```
***Code Snippet***
```
int main(int argc, char** argv)
{
    std::ifstream t("test_buffstring_createdfromimage.txt");
    std::stringstream buffer;
    buffer << t.rdbuf();

    const std::string connectionString=getenv("LOCAL_CONNECTION_STRING");
    //connection string in this case it is "AccountName=localstorageacc;AccountKey=password;BlobEndpoint=http://localhost:11002/localstorageacc"
    const std::string containerName=getenv("LOCAL_CONTAINER_NAME");
    std::string blobName;
    int n=0;

    while(1){

        blobName=std::to_string(n);
        auto start = std::chrono::high_resolution_clock::now();

        auto cClient =  Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(connectionString, containerName);
        auto blobClient = cClient.GetBlockBlobClient(blobName);
        blobClient.UploadFrom(reinterpret_cast<const uint8_t*>(buffer.c_str()), buffer.size());

        auto finish = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> time_elapsed = finish - start;
        std::cout << blobName << ", Elapsed time: " << time_elapsed.count() << " s" << std::endl;
        n=n+1;
    }
}
```
Took reference from this sample code https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/storage/azure-storage-blobs/sample/blob_getting_started.cpp
Buffer stringstream can be any string. The error still occurs when no matter the blob size.

**Expected behavior**
A Elapsed time of 4.5 to 5 seconds will be shown every 100 blobs uploaded.

**Screenshots**
![Screenshot from 2021-07-16 10-09-02](https://user-images.githubusercontent.com/78844564/126034309-831c8f78-ad16-408a-bda1-0450416f97e9.png)

**Setup (please complete the following information):**
 - OS: Ubuntu 20.04 LTS
 - Iot Edge storage module (mcr.microsoft.com/azure-blob-storage:1.4.1-linux-amd64)
 -  Blob storage SDK (azure-sdk-for-cpp, tag: azure-storage-blobs_12.0.0-beta.11-29-g38bd273d)
https://azure.github.io/azure-sdk-for-cpp/, https://github.com/Azure/azure-sdk-for-cpp

**Additional context**
A ticket has been raised under TrackingID#2106230060000825

**Comment by Jinming-Hu** [comment]
@vhvb1989 Looks like an issue in libcurl transport layer. Any thoughts?

**Comment by vhvb1989** [comment]
The delay can be happening either in the transport adapter or as part or the retry policy.

Can we try reproducing with logs enabled? (set AZURE_LOG_LEVEL=1 env var) and then run the application.  We should be able to see exactly where the delay is happening.

**Comment by vhvb1989** [comment]
Also noticed that a beta version is used. (azure-storage-blobs_12.0.0-beta.11)  Shouldn't we use the GA version instead?  Is this reproducible with latest version?

**Comment by vhvb1989** [comment]
Is the application running inside the container?  Or is the Iot Edge storage module only running in the container and your application is using the Azure SDL client from outside the container?
The reason I ask is because the AZURE_LOG_LEVEL should be set in the env where the application is running. The application will output the sdk client logs to `std::cerr`.

Since you mentioned `docker logs`, I wonder if you can explain the full set up.  Also, if you can share your Docker file, I might be able to start the container and try to repro

**Comment by kc-95-scg** [comment]
hi @vhvb1989 

Sorry for the misunderstanding.
I was thinking that we need to expose the variable to the Iotedge blob module (which is running as a docker container). 

I have exported AZURE_LOG_LEVEL=1 and found some good leads.
There seems to be a print out "connection: close"
HTTP Transport error: Connection was closed by the server while trying to read a response
HTTP Retry attempt #1 will be made in 3988ms.

```
[2021-07-19T17:49:01.9835514Z] INFO  : HTTP Response (4ms) : 201 Created
content-length : 0
content-md5 : REDACTED
date : Mon, 19 Jul 2021 17:49:01 GMT
etag : "0x8D94ADD7E7A7BC4"
last-modified : Mon, 19 Jul 2021 17:49:01 GMT
server : Windows-Azure-Blob/1.0
x-ms-request-id : a3120b45-a577-4943-be00-d3f8b8323545
x-ms-request-server-encrypted : REDACTED
x-ms-version : REDACTED
[2021-07-19T17:49:01.9836713Z] WARN  : HTTP status code 201 won't be retried.
200, Elapsed time: 0.00593913 s
[2021-07-19T17:49:01.9843296Z] INFO  : HTTP Request : PUT http://localhost:11002/cgelocaledgeacc/demo/201
authorization : REDACTED
content-length : 0
user-agent : azsdk-cpp-storage-blobs/12.0.1 (Linux 5.8.0-50-generic x86_64 #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021)
x-ms-blob-type : REDACTED
x-ms-client-request-id : 07866d05-9791-4f3d-68d1-557707095a06
x-ms-date : REDACTED
x-ms-version : REDACTED
[2021-07-19T17:49:01.9843934Z] DEBUG : [CURL Transport Adapter]: Creating a new session.
[2021-07-19T17:49:01.9844521Z] DEBUG : [CURL Transport Adapter]: No Host in request headers. Adding it
[2021-07-19T17:49:01.9845257Z] DEBUG : [CURL Transport Adapter]: Using 100-continue for PUT request
[2021-07-19T17:49:01.9845901Z] DEBUG : [CURL Transport Adapter]: Send request without payload
[2021-07-19T17:49:01.9847727Z] DEBUG : [CURL Transport Adapter]: Parse server response
[2021-07-19T17:49:01.9856299Z] DEBUG : [CURL Transport Adapter]: Check server response before upload starts
[2021-07-19T17:49:01.9857212Z] DEBUG : [CURL Transport Adapter]: Upload payload
[2021-07-19T17:49:01.9858842Z] DEBUG : [CURL Transport Adapter]: Upload completed. Parse server response
[2021-07-19T17:49:01.9887774Z] DEBUG : [CURL Transport Adapter]: Request completed. Moving response out of session and session to response.
[2021-07-19T17:49:01.9888842Z] DEBUG : Moving connection to pool...
[2021-07-19T17:49:01.9889658Z] DEBUG : Clean thread running. Won't start a new one.
[2021-07-19T17:49:01.9890853Z] INFO  : HTTP Response (4ms) : 201 Created
connection : close
content-length : 0
content-md5 : REDACTED
date : Mon, 19 Jul 2021 17:49:01 GMT
etag : "0x8D94ADD7E7B543C"
last-modified : Mon, 19 Jul 2021 17:49:01 GMT
server : Windows-Azure-Blob/1.0
x-ms-request-id : 0eec063f-ca52-4086-8195-951f12e4efa7
x-ms-request-server-encrypted : REDACTED
x-ms-version : REDACTED
[2021-07-19T17:49:01.9892880Z] WARN  : HTTP status code 201 won't be retried.
201, Elapsed time: 0.00561808 s
[2021-07-19T17:49:01.9901365Z] INFO  : HTTP Request : PUT http://localhost:11002/cgelocaledgeacc/demo/202
authorization : REDACTED
content-length : 0
user-agent : azsdk-cpp-storage-blobs/12.0.1 (Linux 5.8.0-50-generic x86_64 #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021)
x-ms-blob-type : REDACTED
x-ms-client-request-id : 032270b1-8aa3-4222-553e-eb1f0057624e
x-ms-date : REDACTED
x-ms-version : REDACTED
[2021-07-19T17:49:01.9902360Z] DEBUG : [CURL Transport Adapter]: Creating a new session.
[2021-07-19T17:49:01.9903257Z] DEBUG : [CURL Transport Adapter]: No Host in request headers. Adding it
[2021-07-19T17:49:01.9903819Z] DEBUG : [CURL Transport Adapter]: Using 100-continue for PUT request
[2021-07-19T17:49:01.9904418Z] DEBUG : [CURL Transport Adapter]: Send request without payload
[2021-07-19T17:49:01.9905870Z] DEBUG : [CURL Transport Adapter]: Parse server response
[2021-07-19T17:49:01.9909872Z] WARN  : HTTP Transport error: Connection was closed by the server while trying to read a response
[2021-07-19T17:49:01.9910673Z] INFO  : HTTP Retry attempt #1 will be made in 3988ms.
```

Some additonal information, not sure if it is helpful.
When i am calling the CreateIfNotExists() function for a container client, gdb shows another thread created.
And if i call CreateIfNotExists() too frequently i get a runtime error about XML parsing error.
Happens randomly and not able to reproduce.

**Comment by Jinming-Hu** [comment]
```
I have exported AZURE_LOG_LEVEL=1 and found some good leads.
There seems to be a print out "connection: close"
HTTP Transport error: Connection was closed by the server while trying to read a response
HTTP Retry attempt #1 will be made in 3988ms.
```

So the delay was caused by reusing closed connection.

> And if i call CreateIfNotExists() too frequently i get a runtime error about XML parsing error.

This is another bug that needs to be handled by storage team. Can you open a separate issue and brief the repro steps?

**Comment by vhvb1989** [comment]
If the server is closing the connection, there's nothing the client can do about it.

@kc-95-scg you can use the clientOptions to set the retryOptions. This is one way how you would not wait too much if a retry is required.

Another option is to use the `TransportOptions` to set a transportAdapter which is set as `keepAlive = false`.  This will impact all requests' performance (as each request will create a new connection) but will avoid the reported delay.

Anyway, these are just 2 suggested work-arounds. The right thing to do would be to understand why the Iot Edge Blob Storage Module is closing a connection?  Would it have a max-time?  or does it need something to make sure the connection is not close?

@Jinming-Hu do we know someone from the `Iot Edge Blob Storage Module` who can help to better understand what's happening on the server side? We are blocked on the SDK client side if we don't know more from this Server behavior.
As @kc-95-scg  mentioned, this won't never happen using Azure Storage directly.

Let me know if you want some more info on how to set up/configure the work-arounds I shared (in case it might be interesting for you to try).  

I will update this issue as addressed for now since the investigation should continue on the server side for now.

**Comment by Jinming-Hu** [comment]
> If the server is closing the connection, there's nothing the client can do about it.

Client can destruct the session instead of adding it back to the connection pool.

**Comment by Jinming-Hu** [comment]
> Anyway, these are just 2 suggested work-arounds.

Neither of the two workarounds looks good to me. The first workaround will cause some request to delay for a few seconds. The second workaround will decrease overall performance.

> The right thing to do would be to understand why the Iot Edge Blob Storage Module is closing a connection? Would it have a max-time? or does it need something to make sure the connection is not close?

Closing connection after handling an HTTP request is well-defined behavior. IoT Edge didn't do anything wrong.

> As @kc-95-scg mentioned, this won't never happen using Azure Storage directly.

It's hasty to say **never**. I need to confirm with storage service team.

**Comment by kc-95-scg** [comment]
@vhvb1989 and @Jinming-Hu thanks for the help. We will try with the work around that you suggested and choose the better option. But we would need to raise the urgency as it is affecting our production requirements.
We have been trying to get in contact with Iot Edge Blob Storage Module team, however it seems to be a niche request. Do let us know if you are able to get in contact with someone there.

**Comment by vhvb1989** [comment]
@RickWinter can you help us to find someone to help from the Iot Edge Blob Storage Module?

@Jinming-Hu , when working with SSL, once the secured channel is open to the server, the channel is usually expected to be used as long as there are any communication to happen between the client and server (not only one request).  Opening the SSL channel is time expensive.
For the Azure Storage and other services (like Key Vault), once a secured connection is open to the server, it will be used to complete all the required requests.
From the client perspective, a secured channel (connection) will be discard only if it has more than a minute of inactivity (no data transfer to the server) because a server would typically terminate the session for inactivity after some time (defined by each service).
But, if the client is constantly uploading/downloading data to/from the service, the service is not expected to close the session

@kc-95-scg We will try to find someone from the Iot Edge Blob Storage Module.  For now, you can control the retry-time interval to be shorter.

**Comment by Jinming-Hu** [comment]
Hi @vhvb1989, I've confirmed with FE experts. Storage service will also close connections under certain circumstances. So our HTTP client has to support `Connection:` header defined in [RFC 2616, Section 14.10](https://datatracker.ietf.org/doc/html/rfc2616#section-14.10).

**Comment by vhvb1989** [comment]
>  Storage service will also close connections under certain circumstances.

That's expected, of course. There are many reasons why a Server would decide to close a connection. For example, a bad request is received, or an error response is returned to the client, etc. 
I am not saying that the server should never close a connection.  What I say is that if  there is not any `certain circumstance`, the server should not close the connection.  And this current issue looks like such a case. A client is just constantly uploading a blob, over and over (same request is performed in a loop).  
So, the server won't close the connection due to a bad request (as all are the same).

Since the closed-connection is always happening after 100 uploads, I could guess that the server is closing the connection due to a time-limit constraint or data-transfer limit, or something like that.  That's what we need to know before we can try any change for the client.

> So our HTTP client has to support Connection: header defined

This header is supported/implemented by a server.  It is similar to `Connection: Keep-Alive`.  The client would just send the header and expect the server to use it as an instruction for the request.  @Jinming-Hu what do  you mean here?  Are you suggesting that the SDK client should always use this header?

**Comment by Jinming-Hu** [comment]
> That's expected, of course. There are many reasons why a Server would decide to close a connection. For example, a bad request is received, or an error response is returned to the client, etc.
I am not saying that the server should never close a connection. What I say is that if there is not any certain circumstance, the server should not close the connection. And this current issue looks like such a case. A client is just constantly uploading a blob, over and over (same request is performed in a loop).
So, the server won't close the connection due to a bad request (as all are the same).

This is not true. Storage service may close the any connection at any time. Even if you constantly upload blobs, all of the requests are successful, storage service may close the connection after some of the requests.

```
   HTTP/1.1 defines the "close" connection option for the sender to
   signal that the connection will be closed after completion of the
   response. For example,

       Connection: close

   in either the request or the response header fields indicates that
   the connection SHOULD NOT be considered `persistent' (section 8.1)
   after the current request/response is complete.
```

This header is also returned by the server. If HTTP client receives such a header, it should close this connection instead of adding it back to the connection pool.

**Comment by vhvb1989** [comment]
A `sender` refers to a client.  in HTTP, there's always a sender and a receiver.  

Anyway, I can see in the logs that the server is actually returning that header `connection : close` .  I will update our curl-transport adapter implementation for such a case.  

I have updated this issue and I expect to have the fix included in the next month release (August 2021). Thank you @Jinming-Hu

**Comment by kc-95-scg** [comment]
@vhvb1989 thanks for the action and sorry for the late replies.
Just had the time to test out your 2 suggested work arounds, while waiting for the fix.
However, I am not able to find the implementation methods. (Could be my mistake as I am not very proficient with the SDK)
1) Regarding the clientOptions, are you refering to Azure::Storage::Blobs::BlobClientOptions?
I dont see a retryOptions member.

2) Regarding the keepAlive = false for TransportOptions. 
May I know if you are referring to Azure::Storage::Blobs::UploadBlockBlobFromOptions?
I dont see the member "keepAlive" in the TransportOptions struct as well

Possible to give an example?
ie. something like this?
```
Azure::Storage::Blobs::BlobClientOptions clientoptions;
clientoptions.retryOptions = ??;
auto cClient =  Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString (connectionString, blobContainerName, clientoptions);
```

```
Azure::Storage::Blobs::UploadBlockBlobFromOptions bloboptions;
bloboptions.TransferOptions.keepAlive=false;
blobClient.UploadFrom(reinterpret_cast<const uint8_t*>(buffer.c_str()), buffer.size(),bloboptions); 
```

**Comment by kc-95-scg** [comment]
Dear @vhvb1989 and @Jinming-Hu possible to help us out with the workaround solution above?

**Comment by vhvb1989** [comment]
@kc-95-scg yes, my apologies for the late reply.

So, let me take the source code you initially provided and show you how you can set the client up to and try this work arounds.

## Using retry options:

Updating your src code below to set the Retry Options for the client. I added a comment with some notes.
If you want to learn all you can do with the retry policy, take a look to: https://azure.github.io/azure-sdk/general_azurecore.html#retry-policy

### Notes:
- Setting the retry-delay to a short time could bring a side effect. For example, if the computer where the application is running loses the network connection during a period time of 10 seconds and the retry options are set up to make 3 retry intents with a delay of some milliseconds, an operation would fail as well as all the retry intendents.  But if the retry-delay is set to a longer period of time (like 4 or more seconds), the chances from succeeding an operation, even from a network interruption, are some how better. So, the retry options is a tool for customers to be set up to their best convenience.

```cpp
int main(int argc, char** argv)
{
  std::ifstream t("test_buffstring_createdfromimage.txt");
  std::stringstream buffer;
  buffer << t.rdbuf();

  const std::string connectionString = getenv("LOCAL_CONNECTION_STRING");
  // connection string in this case it is
  // "AccountName=localstorageacc;AccountKey=password;BlobEndpoint=http://localhost:11002/localstorageacc"
  const std::string containerName = getenv("LOCAL_CONTAINER_NAME");
  std::string blobName;
  int n = 0;

  while (1)
  {

    blobName = std::to_string(n);
    auto start = std::chrono::high_resolution_clock::now();

    // Use client options before creating the client and use it as the last optional arg when
    // creating the client:
    Azure::Storage::Blobs::BlobClientOptions options;
    options.Retry.RetryDelay = std::chrono::milliseconds(1);
    auto cClient = Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
        connectionString, containerName, options);

    // At this point, all the operations from `cClient` will use the specified retry-configuration.
    // If you want to learn more about the retry policy see:
    // https://azure.github.io/azure-sdk/general_azurecore.html#retry-policy
    auto blobClient = cClient.GetBlockBlobClient(blobName);
    blobClient.UploadFrom(reinterpret_cast<const uint8_t*>(buffer.c_str()), buffer.size());

    auto finish = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> time_elapsed = finish - start;
    std::cout << blobName << ", Elapsed time: " << time_elapsed.count() << " s" << std::endl;
    n = n + 1;
}
```

## Avoid re-using HTTP connections

In this scenario, there is also a side effect.  The performance from the client to complete the operations is impacted in the scale of 1/3. The SDK client will take up to 1 extra second per each request. This performance penalty is even more visible if the client is performing multiple small operations (upload/download small some bytes only) instead of a long operation (upload/download MB/GB).

```cpp
// We need to include the curl transport adapter header to customize it different from its default behavior
#include <azure/core/http/curl_transport.hpp>

int main(int argc, char** argv)
{
  std::ifstream t("test_buffstring_createdfromimage.txt");
  std::stringstream buffer;
  buffer << t.rdbuf();

  const std::string connectionString = getenv("LOCAL_CONNECTION_STRING");
  // connection string in this case it is
  // "AccountName=localstorageacc;AccountKey=password;BlobEndpoint=http://localhost:11002/localstorageacc"
  const std::string containerName = getenv("LOCAL_CONTAINER_NAME");
  std::string blobName;
  int n = 0;

  while (1)
  {

    blobName = std::to_string(n);
    auto start = std::chrono::high_resolution_clock::now();

    // We need to use a custom set-up transport adapter. Then set this adapter in the client
    // options.
    Azure::Core::Http::CurlTransportOptions curlAdapterOptions;
    // Make transport to use one connection per request
    curlAdapterOptions.HttpKeepAlive = false;
    Azure::Storage::Blobs::BlobClientOptions options;
    options.Transport.Transport = std::make_unique<Azure::Core::Http::CurlTransport>(curlAdapterOptions);
    auto cClient = Azure::Storage::Blobs::BlobContainerClient::CreateFromConnectionString(
        connectionString, containerName, options);

    // At this point, all the operations from `cClient` will use the custom transport adapter.
    // If you want to learn more about the transport adapter see:
    // https://github.com/Azure/azure-sdk-for-cpp/blob/main/doc/HttpTransportAdapter.md
    auto blobClient = cClient.GetBlockBlobClient(blobName);
    blobClient.UploadFrom(reinterpret_cast<const uint8_t*>(buffer.c_str()), buffer.size());

    auto finish = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> time_elapsed = finish - start;
    std::cout << blobName << ", Elapsed time: " << time_elapsed.count() << " s" << std::endl;
    n = n + 1;
}
```

@kc-95-scg  I am working on a fix for you for the client to close the connection upon getting the `connection: close` header from the server.  You can give a try to any of these workarounds and see if it works for you.  Then, once the fix is merged and released, you might be able to remove the work-around and just use your original source code.

Let me know if there's any other thing I can help you with.

**Comment by kc-95-scg** [comment]
Dear @vhvb1989 thanks for the extensive reply :)
We have tested the workaround using the HttpKeepAlive = false and we don't find a significant loss in performance. We will go ahead with this solution. Thank you very much.
